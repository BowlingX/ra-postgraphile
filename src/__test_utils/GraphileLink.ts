import {
  ApolloLink,
  Operation,
  NextLink,
  FetchResult,
  Observable,
  ExecutionResult,
} from 'apollo-link'
import { execute, getOperationAST, GraphQLSchema } from 'graphql'
import { Pool } from 'pg'
import { withPostGraphileContext } from 'postgraphile'

interface GraphileLinkInterface {
  pgPool: Pool
  schema: GraphQLSchema
}

export default class GraphileLink extends ApolloLink {
  constructor(private options: GraphileLinkInterface) {
    super()
  }

  request(
    operation: Operation,
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    _forward?: NextLink
  ): Observable<FetchResult> | null {
    const { schema, pgPool } = this.options
    return new Observable<any>((observer) => {
      // eslint-disable-next-line no-extra-semi
      ;(async () => {
        try {
          const op = getOperationAST(operation.query, operation.operationName)
          /* Only do queries (not subscriptions)  */
          if (!op || op.operation === 'subscription') {
            if (!observer.closed) {
              observer.complete()
            }
            return
          }

          const data = await withPostGraphileContext(
            { pgPool },
            (context) =>
              execute(
                schema,
                operation.query,
                {},
                context,
                operation.variables,
                operation.operationName
              ) as Promise<ExecutionResult>
          )
          if (!observer.closed) {
            observer.next(data)
            observer.complete()
          }
        } catch (e) {
          if (!observer.closed) {
            observer.error(e)
          } else {
            console.error(e)
          }
        }
      })()
    })
  }
}
