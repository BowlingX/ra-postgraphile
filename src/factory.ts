import buildGraphQLProvider, { Options } from 'ra-data-graphql'
import type { ApolloClient } from '@apollo/client'
import { buildQuery } from './buildQuery'
import { buildInTypeConfig } from './defaultTypeConfig'
import { ProviderOptions } from './types'

export const factory = <T = any>(
  client: ApolloClient<T>,
  options?: ProviderOptions,
  graphqlProviderOptions?: Omit<Options, 'client' | 'buildQuery'>
): ReturnType<typeof buildGraphQLProvider> => {
  const defaultAppliedOptions = {
    typeMap: {
      ...buildInTypeConfig,
      ...(options?.typeMap || {}),
    },
  }

  return buildGraphQLProvider({
    ...graphqlProviderOptions,
    client,
    buildQuery: buildQuery(defaultAppliedOptions),
  })
}
