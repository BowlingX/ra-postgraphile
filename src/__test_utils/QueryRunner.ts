import 'dotenv/config'
import { introspectionFromSchema } from 'graphql'
import { ApolloClient, InMemoryCache } from '@apollo/client'
import { Pool } from 'pg'
import { createPostGraphileSchema } from 'postgraphile'

import PgSimplifyInflectorPlugin from '@graphile-contrib/pg-simplify-inflector'
// @ts-ignore
import PgConnectionFilterPlugin from 'postgraphile-plugin-connection-filter'
import GraphileLink from './GraphileLink'

export async function makeQueryRunner(
  connectionString = process.env.DATABASE_URL || 'postgres:///',
  schemaName = process.env.DATABASE_SCHEMA || 'app_public',
  options = {
    appendPlugins: [PgSimplifyInflectorPlugin, PgConnectionFilterPlugin],
  } // See https://www.graphile.org/postgraphile/usage-schema/ for options
) {
  const pgPool = new Pool({
    connectionString,
  })

  // Create the PostGraphile schema
  const schema = await createPostGraphileSchema(connectionString, schemaName, options)
  const introspection = introspectionFromSchema(schema, { descriptions: true })

  const apolloClient = new ApolloClient({
    ssrMode: true,
    cache: new InMemoryCache(),
    link: new GraphileLink({ pgPool, schema }),
  })

  const release = () => {
    pgPool.end()
  }

  return {
    release,
    apolloClient,
    schema: introspection.__schema,
  }
}
