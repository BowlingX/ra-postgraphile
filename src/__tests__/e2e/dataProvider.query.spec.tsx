import expect from 'expect'
import { ApolloClient } from 'apollo-client'
import { convertLegacyDataProvider, DataProvider } from 'ra-core'
import { makeQueryRunner } from '../../__test_utils/QueryRunner'
import { factory } from '../../factory'

let client: ApolloClient<any>
let cleanup: () => void
let dataProvider: DataProvider

beforeAll(async () => {
  const { release, apolloClient, schema } = await makeQueryRunner()
  cleanup = release
  client = apolloClient
  const legacyProvider = await factory(client, undefined, { introspection: { schema } })
  dataProvider = convertLegacyDataProvider(legacyProvider)
})

afterAll(() => {
  cleanup()
})

describe('dataProvider', () => {
  it('should construct from a schema', async () => {
    expect(dataProvider).toBeDefined()
  })
})

describe('Primary keys', () => {
  it('should handle non-id field primary keys with `nodeId`', async () => {
    const favoritesList = await dataProvider.getList('favoriteBook', {
      sort: { field: 'isbn', order: 'ASC' },
      filter: {},
      pagination: { perPage: 10, page: 1 },
    })
    expect(favoritesList).toMatchSnapshot()
  })

  it('should fetch a single item by nodeId', async () => {
    const favoritesList = await dataProvider.getList('favoriteBook', {
      sort: { field: 'isbn', order: 'ASC' },
      filter: {},
      pagination: { perPage: 10, page: 1 },
    })

    const singleBook = await dataProvider.getOne('favoriteBook', {
      id: favoritesList?.data[0]?.id,
    })
    expect(singleBook?.data?.isbn).toEqual('3221123')
  })

  it('should fetch many items by nodeId', async () => {
    const favoritesList = await dataProvider.getList('favoriteBook', {
      sort: { field: 'isbn', order: 'ASC' },
      filter: {},
      pagination: { perPage: 10, page: 1 },
    })

    const manyBooks = await dataProvider.getMany('favoriteBook', {
      ids: favoritesList?.data.map((data) => data.isbn),
    })
    console.log(manyBooks)
  })

  it('should be able to delete for non-id field primary keys with `deleteByNodeId`', async () => {
    const data = await dataProvider.getList('favoriteBook', {
      pagination: { perPage: 10, page: 1 },
      filter: {
        isbn: '3221123',
      },
      sort: { field: 'isbn', order: 'ASC' },
    })
    const deletedBook = await dataProvider.delete('favoriteBook', {
      id: data?.data[0].id,
      previousData: data?.data[0],
    })
    console.log(deletedBook)
  })
})

describe('Query Types', () => {
  describe('should getOne', () => {
    it('should find a book', async () => {
      expect(await dataProvider.getOne('book', { id: 1 })).toMatchSnapshot()
    })
    it('should return null if book was not found', async () => {
      expect(await dataProvider.getOne('book', { id: 5 })).toMatchSnapshot()
    })
  })

  describe('should getList', () => {
    it('should list all books without filter and sorting', async () => {
      expect(
        await dataProvider.getList('book', {
          // @ts-ignore
          sort: {},
          filter: {},
          pagination: { perPage: 10, page: 1 },
        })
      ).toMatchSnapshot()
    })

    it('should list all books with sorting', async () => {
      expect(
        await dataProvider.getList('book', {
          sort: { field: 'id', order: 'ASC' },
          filter: {},
          pagination: { perPage: 10, page: 1 },
        })
      ).toMatchSnapshot()
    })

    it('should filter books', async () => {
      expect(
        await dataProvider.getList('book', {
          sort: { order: 'ASC', field: 'id' },
          filter: { id: 1 },
          pagination: { perPage: 10, page: 1 },
        })
      ).toMatchSnapshot()
    })

    it('should use `$contains` to filter books', async () => {
      const books = await dataProvider.getList('book', {
        sort: { order: 'ASC', field: 'id' },
        filter: { $condition: { id: 1 } },
        pagination: { perPage: 10, page: 1 },
      })
      expect(books?.data?.length === 1)
    })

    describe('View', () => {
      it('should work with views', async () => {
        expect(
          await dataProvider.getList('allFavoriteBooks', {
            sort: { field: 'id', order: 'ASC' },
            filter: {},
            pagination: { perPage: 10, page: 1 },
          })
        ).toMatchSnapshot()
      })
    })

    describe('Function', () => {
      it('should be able to fetch a resource by a `setof` function and custom type', async () => {
        expect(
          await dataProvider.getList('MyCustomBook', {
            sort: { field: 'id', order: 'ASC' },
            filter: {},
            pagination: { perPage: 10, page: 1 },
          })
        ).toMatchSnapshot()
      })
    })

    describe('Error resilience', () => {
      it('should fail on unknown resources', async () => {
        await expect(dataProvider.getOne('unknownResource', { id: 1 })).rejects.toEqual(
          Error('Type "UnknownResource" did not exist in the introspection result.')
        )
      })

      it('should allow lowercase fields to sort on', async () => {
        expect(
          await dataProvider.getList('book', {
            sort: { order: 'asc', field: 'id' },
            filter: {},
            pagination: { perPage: 10, page: 1 },
          })
        ).toMatchSnapshot()
      })

      it('should work on uppercase resource names', async () => {
        expect(await dataProvider.getOne('Book', { id: 1 })).toMatchSnapshot()
      })
    })
  })

  describe('UUID type (#48)', () => {
    it('should list `Profile`', async () => {
      const result = await dataProvider.getList('profile', {
        sort: { order: 'asc', field: 'id' },
        filter: {},
        pagination: { perPage: 10, page: 1 },
      })
      expect(result.data[0]).toBeDefined()
      expect(result.data[0]?.name).toEqual('David')
    })
  })
})
