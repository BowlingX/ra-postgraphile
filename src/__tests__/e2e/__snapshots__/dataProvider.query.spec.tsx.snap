// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Non standard Primary keys should be able to delete\` 1`] = `
Object {
  "data": Object {
    "__typename": "FavoriteBook",
    "id": "3221123",
    "isSuperFavorite": false,
    "isbn": "3221123",
    "nodeId": "WyJmYXZvcml0ZV9ib29rcyIsIjMyMjExMjMiXQ==",
  },
}
`;

exports[`Non standard Primary keys should be able to update many 1`] = `
Object {
  "data": Array [
    "3221123",
    "51231",
  ],
}
`;

exports[`Non standard Primary keys should be able to update many 2`] = `
Object {
  "data": Array [
    Object {
      "__typename": "FavoriteBook",
      "id": "3221123",
      "isSuperFavorite": true,
      "isbn": "3221123",
      "nodeId": "WyJmYXZvcml0ZV9ib29rcyIsIjMyMjExMjMiXQ==",
    },
    Object {
      "__typename": "FavoriteBook",
      "id": "51231",
      "isSuperFavorite": true,
      "isbn": "51231",
      "nodeId": "WyJmYXZvcml0ZV9ib29rcyIsIjUxMjMxIl0=",
    },
  ],
  "total": 2,
}
`;

exports[`Non standard Primary keys should fetch many items 1`] = `
Object {
  "data": Array [
    Object {
      "__typename": "FavoriteBook",
      "id": "3221123",
      "isSuperFavorite": false,
      "isbn": "3221123",
      "nodeId": "WyJmYXZvcml0ZV9ib29rcyIsIjMyMjExMjMiXQ==",
    },
    Object {
      "__typename": "FavoriteBook",
      "id": "51231",
      "isSuperFavorite": false,
      "isbn": "51231",
      "nodeId": "WyJmYXZvcml0ZV9ib29rcyIsIjUxMjMxIl0=",
    },
  ],
}
`;

exports[`Non standard Primary keys should list them 1`] = `
Object {
  "data": Array [
    Object {
      "__typename": "FavoriteBook",
      "id": "3221123",
      "isSuperFavorite": false,
      "isbn": "3221123",
      "nodeId": "WyJmYXZvcml0ZV9ib29rcyIsIjMyMjExMjMiXQ==",
    },
    Object {
      "__typename": "FavoriteBook",
      "id": "51231",
      "isSuperFavorite": false,
      "isbn": "51231",
      "nodeId": "WyJmYXZvcml0ZV9ib29rcyIsIjUxMjMxIl0=",
    },
  ],
  "total": 2,
}
`;

exports[`Query Types should getList Error resilience should allow lowercase fields to sort on 1`] = `
Object {
  "data": Array [
    Object {
      "__typename": "Book",
      "authorId": 1,
      "id": 1,
      "isbn": "12345",
      "name": "Moby Dick or The Whale",
      "nodeId": "WyJib29rcyIsMV0=",
    },
    Object {
      "__typename": "Book",
      "authorId": 2,
      "id": 2,
      "isbn": "51231",
      "name": "The China Study",
      "nodeId": "WyJib29rcyIsMl0=",
    },
    Object {
      "__typename": "Book",
      "authorId": 3,
      "id": 3,
      "isbn": "3221123",
      "name": "How Not to Die",
      "nodeId": "WyJib29rcyIsM10=",
    },
  ],
  "total": 3,
}
`;

exports[`Query Types should getList Error resilience should work on uppercase resource names 1`] = `
Object {
  "data": Object {
    "__typename": "Book",
    "authorId": 1,
    "id": 1,
    "isbn": "12345",
    "name": "Moby Dick or The Whale",
    "nodeId": "WyJib29rcyIsMV0=",
  },
}
`;

exports[`Query Types should getList Function should be able to fetch a resource by a \`setof\` function and custom type 1`] = `
Object {
  "data": Array [
    Object {
      "__typename": "MyCustomBook",
      "id": 2,
      "name": "The China Study",
    },
  ],
  "total": 1,
}
`;

exports[`Query Types should getList View should work with views 1`] = `
Object {
  "data": Array [
    Object {
      "__typename": "AllFavoriteBook",
      "id": 2,
      "name": "The China Study",
    },
    Object {
      "__typename": "AllFavoriteBook",
      "id": 3,
      "name": "How Not to Die",
    },
  ],
  "total": 2,
}
`;

exports[`Query Types should getList should filter books 1`] = `
Object {
  "data": Array [
    Object {
      "__typename": "Book",
      "authorId": 1,
      "id": 1,
      "isbn": "12345",
      "name": "Moby Dick or The Whale",
      "nodeId": "WyJib29rcyIsMV0=",
    },
  ],
  "total": 1,
}
`;

exports[`Query Types should getList should list all books with sorting 1`] = `
Object {
  "data": Array [
    Object {
      "__typename": "Book",
      "authorId": 1,
      "id": 1,
      "isbn": "12345",
      "name": "Moby Dick or The Whale",
      "nodeId": "WyJib29rcyIsMV0=",
    },
    Object {
      "__typename": "Book",
      "authorId": 2,
      "id": 2,
      "isbn": "51231",
      "name": "The China Study",
      "nodeId": "WyJib29rcyIsMl0=",
    },
    Object {
      "__typename": "Book",
      "authorId": 3,
      "id": 3,
      "isbn": "3221123",
      "name": "How Not to Die",
      "nodeId": "WyJib29rcyIsM10=",
    },
  ],
  "total": 3,
}
`;

exports[`Query Types should getList should list all books without filter and sorting 1`] = `
Object {
  "data": Array [
    Object {
      "__typename": "Book",
      "authorId": 1,
      "id": 1,
      "isbn": "12345",
      "name": "Moby Dick or The Whale",
      "nodeId": "WyJib29rcyIsMV0=",
    },
    Object {
      "__typename": "Book",
      "authorId": 2,
      "id": 2,
      "isbn": "51231",
      "name": "The China Study",
      "nodeId": "WyJib29rcyIsMl0=",
    },
    Object {
      "__typename": "Book",
      "authorId": 3,
      "id": 3,
      "isbn": "3221123",
      "name": "How Not to Die",
      "nodeId": "WyJib29rcyIsM10=",
    },
  ],
  "total": 3,
}
`;

exports[`Query Types should getOne should find a book 1`] = `
Object {
  "data": Object {
    "__typename": "Book",
    "authorId": 1,
    "id": 1,
    "isbn": "12345",
    "name": "Moby Dick or The Whale",
    "nodeId": "WyJib29rcyIsMV0=",
  },
}
`;

exports[`Query Types should getOne should return null if book was not found 1`] = `
Object {
  "data": null,
}
`;
